lib/one_gadget/gadget.rb

```diff
@@ -81,6 +83,7 @@ def calculate_score(expr)
         when /^writable/ then calculate_writable_score(expr.sub('writable: ', ''))
         when / == NULL$/ then calculate_null_score(expr.sub(' == NULL', ''))
         when / <= 0$/ then calculate_null_score(expr.sub(' <= 0', ''))
+        when / is a valid (argv|envp)$/ then 0.2 # TODO: Calculate the score more precisely
```

Comment:
I feel we can give it score 1, since it basically means something on stack is NULL (same as the case "[sp+xx] == NULL is easy" in `calculate_null_score`)

Reply:
I have some reservations about assigning a score of 1.
Is it truly correct to attribute such a high score to it? This is particularly relevant when we examine cases like `{"sh", "-c", x, NULL}`.
In such instances, the value of x must hold a readable and contextually meaningful address.
However, manipulating x to an address that meets these criteria might not easy to be feasible most of the time (?).
Furthermore, when addressing cases like `{"/bin/sh", x, NULL}`, the score attributed to this constraint should probably equal to or less than the score derived from `calculate_null_score(x)`, instead of always 1.

To address the scenario I mentioned earlier, my initial approach involved creating a set of comprehensive conditions, such as `"[#{arg}] == NULL || ... is a valid argv"`.
By assigning a relatively low score of 0.2 to the constraint for `is valid argv`, the dominant influence within the entire `||` constraint becomes `[#{arg}] == NULL` for the majority of cases.
The `is valid argv` constraint essentially serves as supplementary information and acts as a fallback solution. Its significance truly emerges when our attempts to satisfy all other constraints have proven unsuccessful.

Nevertheless, my approach has given rise to an additional predicament, as can be gleaned from this instance: https://github.com/david942j/one_gadget/pull/206#discussion_r1277539183.

So, I think we have the following options:
- Simplify the constraints when we generated it, i.e. directly fixing somewhere like this: https://github.com/david942j/one_gadget/pull/206#discussion_r1277539183, and re-parsing the `...` in `... is a valid argv/envp` for calculating the more accurate score here.
- Keep this for now, until we want more accurate score :). (I personally prefer this option haha. Even if we make the score more accurate, it probably doesn't have a difference in the same output level. In addition, I found the verbose constraints like `[r12] == NULL || r12 is a valid envp` to be more intuitive than a single `r12 is a valid envp`, although `[r12] == NULL` is redundancy.)

What do you think?

Reply:
Thanks for the explanation and I am convinced it's different from the NULL cases indeed. Let's go with current implementation and also remove `TODO: Calculate the score more precisely`, as you said I don't think it's worthy to to enhance it. Instead, add a comment that "/ is a valid (argv|envp)$/" usually means the register has to be a readable and pointer.

Reply:
Ok, no problem!